### ISSUE TITLE
[object_store] Object Spilling does not work properly

### ISSUE THREAD

**Issue by franklsf95** [ray-project/ray#14788]
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*

* Ray: nightly (3/18/2021) via https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-2.0.0.dev0-cp38-cp38-manylinux2014_x86_64.whl
* Python 3.8.5
* Linux ip-172-31-61-153 5.4.0-1038-aws #40~18.04.1-Ubuntu SMP Sat Feb 6 01:56:56 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

I ran this on a m5.xlarge instance with 4 CPU and 16GB RAM.

First, start Ray with the following command:

```
ray stop && ray start --head --system-config='{"automatic_object_spilling_enabled":true,"max_io_workers":4,"min_spilling_size":104857600,"object_spilling_config":"{\"type\":\"filesystem\",\"params\":{\"directory_path\":\"/tmp/spill\"}}"}' --resources='{"worker":1}'
```

Second, run the repro script (see end):

Observe:

* The two mappers start, printing "Mapper 0" and "Mapper 1"
* Objects seem to be spilled onto disk as expected
* The system then hangs

If you change the `BUF_SIZE` to be something smaller (e.g. `1e9`), then the program finishes normally.

cc @stephanie-wang @rkooo567 

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

----

Repro script code (edited):

```python
import numpy as np
import ray

MAPPER_PART_SIZE = 2 * 1024 * 1024 * 1024
M = 4
R = 4


@ray.remote(resources={"worker": 1})
def mapper(mapper_id):
    print("Mapper", mapper_id)
    part_size = int(MAPPER_PART_SIZE / R)
    ret = [b"0" * part_size for _ in range(R)]
    return ret


@ray.remote
def reducer(reducer_id, *chunks):
    print("Reducer", reducer_id)
    print([len(c) for c in chunks])
    return 0


ray.init(address="auto")

mapper_results = np.empty((M, R), dtype=object)

for m in range(M):
    mapper_results[m, :] = mapper.options(num_returns=R).remote(m)

reducer_results = []
for r in range(R):
    chunks = mapper_results[:, r].tolist()
    ret = reducer.remote(r, *chunks)
    reducer_results.append(ret)

ray.get(reducer_results)
print("OK")
```

**Comment by ericl** [comment]
@franklsf95 can you take a deeper look into what's going on here?

**Comment by franklsf95** [comment]
It turns out the worker processes crashed because of `print(chunks)` which tries to print 8GB worth of bytes. Removing that line fixes the problem.

However Ray does print a warning which I'm not sure if is intended or necessary:

```
(pid=14411) Mapper 0
(pid=14411) Mapper Return 0
(pid=14411) Mapper 1
(pid=14411) Mapper Return 1
2021-03-22 18:39:17,007 WARNING worker.py:1107 -- Traceback (most recent call last):
  File "python/ray/_raylet.pyx", line 661, in ray._raylet.restore_spilled_objects_handler
  File "python/ray/_raylet.pyx", line 664, in ray._raylet.restore_spilled_objects_handler
  File "/home/ubuntu/miniconda3/envs/raysort/lib/python3.8/site-packages/ray/external_storage.py", line 379, in restore_spilled_objects
    return _external_storage.restore_spilled_objects(object_refs,
  File "/home/ubuntu/miniconda3/envs/raysort/lib/python3.8/site-packages/ray/external_storage.py", line 238, in restore_spilled_objects
    self._put_object_to_store(metadata, buf_len, f, object_ref)
  File "/home/ubuntu/miniconda3/envs/raysort/lib/python3.8/site-packages/ray/external_storage.py", line 91, in _put_object_to_store
    worker.core_worker.put_file_like_object(metadata, data_size, file_like,
  File "python/ray/_raylet.pyx", line 957, in ray._raylet.CoreWorker.put_file_like_object
  File "python/ray/_raylet.pyx", line 924, in ray._raylet.CoreWorker._create_put_buffer
  File "python/ray/_raylet.pyx", line 145, in ray._raylet.check_status
ray.exceptions.ObjectStoreFullError: Failed to put object 63964fa4841d4a2effffffffffffffffffffffff0100000002000000 in object store because it is full. Object size is 2147483648 bytes.
The local object store is full of objects that are still in scope and cannot be evicted. Tip: Use the `ray memory` command to list active objects in the cluster.
An unexpected internal error occurred while the IO worker was restoring spilled objects.
(pid=17038) Reducer 0
(pid=17038) [2147483648, 2147483648]
(pid=19273) Reducer 1
(pid=19273) [2147483648, 2147483648]
OK
```

If this is expected behavior, then we can close this.

**Comment by ericl** [comment]
Hmm I think this kind of issue is expected right now, though we have some
fixes to try to avoid this kind of memory overcommit in the common case. To
avoid this, try using smaller chunks (chunks much smaller than the object
store size).

**Comment by rkooo567** [comment]
Btw, I agree that we need a better warning, or this should be more obvious in `ray memory` command (for usability).

**Comment by franklsf95** [comment]
I updated the repro script code. This works well now for 4 mappers, 4 reducers with 2GB per mapper (8GB total data).

The error messages can be improved, but I think we can close this now?

**Comment by franklsf95** [comment]
BTW, when I have `M = R = 4` and 4GB per mapper, Ray consistently crashes with the following error:

There would now be 16 1GB objects in the store. I set the object store memory to be a little over 4GB, so this is stretching it a bit. But for usability I think the error message should be better.

```
(pid=6966) Mapper 0
(pid=6966) Mapper 1
(pid=6966) Mapper 2
(pid=6966) Mapper 3
(raylet) [2021-03-23 10:13:18,161 C 12236 12236] local_object_manager.cc:316:  Check failed: objects_pending_restore_.emplace(object_id).second Object dedupe wasn't done properly. Please report if you see this issue.
(raylet) [2021-03-23 10:13:18,181 E 12236 12236] logging.cc:415: *** Aborted at 1616519598 (unix time) try "date -d @1616519598" if you are using GNU date ***
(raylet) [2021-03-23 10:13:18,188 E 12236 12236] logging.cc:415: PC: @                0x0 (unknown)
(raylet) [2021-03-23 10:13:18,227 E 12236 12236] logging.cc:415: *** SIGABRT (@0x3e800002fcc) received by PID 12236 (TID 0x7f928dc00800) from PID 12236; stack trace: ***
(raylet) [2021-03-23 10:13:18,247 E 12236 12236] logging.cc:415:     @     0x5614c77925df google::(anonymous namespace)::FailureSignalHandler()
(raylet) [2021-03-23 10:13:18,247 E 12236 12236] logging.cc:415:     @     0x7f928d7de980 (unknown)
(raylet) [2021-03-23 10:13:18,247 E 12236 12236] logging.cc:415:     @     0x7f928c887fb7 gsignal
(raylet) [2021-03-23 10:13:18,247 E 12236 12236] logging.cc:415:     @     0x7f928c889921 abort
(raylet) [2021-03-23 10:13:18,252 E 12236 12236] logging.cc:415:     @     0x5614c73483a6 _ZN3ray6RayLogD2Ev.cold
(raylet) [2021-03-23 10:13:18,257 E 12236 12236] logging.cc:415:     @     0x5614c73eb673 _ZZN3ray6raylet18LocalObjectManager25AsyncRestoreSpilledObjectERKNS_8ObjectIDERKSsSt8functionIFvRKNS_6StatusEEEENKUlSt10shared_ptrINS0_15WorkerInterfaceEEE_clESF_
(raylet) [2021-03-23 10:13:18,261 E 12236 12236] logging.cc:415:     @     0x5614c73eb9b5 _ZNSt17_Function_handlerIFvSt10shared_ptrIN3ray6raylet15WorkerInterfaceEEEZNS2_18LocalObjectManager25AsyncRestoreSpilledObjectERKNS1_8ObjectIDERKSsSt8functionIFvRKNS1_6StatusEEEEUlS4_E_E9_M_invokeERKSt9_Any_dataOS4_
(raylet) [2021-03-23 10:13:18,271 E 12236 12236] logging.cc:415:     @     0x5614c73c93bd ray::raylet::WorkerPool::PushIOWorkerInternal()
(raylet) [2021-03-23 10:13:18,271 E 12236 12236] logging.cc:415:     @     0x5614c73c95a8 ray::raylet::WorkerPool::PushRestoreWorker()
(raylet) [2021-03-23 10:13:18,271 E 12236 12236] logging.cc:415:     @     0x5614c73e78fb _ZNSt17_Function_handlerIFvRKN3ray6StatusERKNS0_3rpc26RestoreSpilledObjectsReplyEEZZNS0_6raylet18LocalObjectManager25AsyncRestoreSpilledObjectERKNS0_8ObjectIDERKSsSt8functionIFvS3_EEENKUlSt10shared_ptrINS9_15WorkerInterfaceEEE_clESL_EUlS3_S7_E_E9_M_invokeERKSt9_Any_dataS3_S7_
(raylet) [2021-03-23 10:13:18,276 E 12236 12236] logging.cc:415:     @     0x5614c73f174f ray::rpc::ClientCallImpl<>::OnReplyReceived()
(raylet) [2021-03-23 10:13:18,280 E 12236 12236] logging.cc:415:     @     0x5614c7402c82 _ZN5boost4asio6detail18completion_handlerIZN3ray3rpc17ClientCallManager29PollEventsFromCompletionQueueEiEUlvE_E11do_completeEPvPNS1_19scheduler_operationERKNS_6system10error_codeEm
(raylet) [2021-03-23 10:13:18,301 E 12236 12236] logging.cc:415:     @     0x5614c7affaa1 boost::asio::detail::scheduler::do_run_one()
(raylet) [2021-03-23 10:13:18,302 E 12236 12236] logging.cc:415:     @     0x5614c7b01149 boost::asio::detail::scheduler::run()
(raylet) [2021-03-23 10:13:18,307 E 12236 12236] logging.cc:415:     @     0x5614c7b03637 boost::asio::io_context::run()
(raylet) [2021-03-23 10:13:18,311 E 12236 12236] logging.cc:415:     @     0x5614c7364182 main
(raylet) [2021-03-23 10:13:18,311 E 12236 12236] logging.cc:415:     @     0x7f928c86abf7 __libc_start_main
(raylet) [2021-03-23 10:13:18,313 E 12236 12236] logging.cc:415:     @     0x5614c7378b55 (unknown)
```

**Comment by ericl** [comment]
That's seems bad, is this on latest nightly?

**Comment by franklsf95** [comment]
@ericl yes

**Comment by rkooo567** [comment]
This must be a bug. We should fix it

**Comment by ericl** [comment]
I think this is likely fixed by https://github.com/ray-project/ray/pull/14821 which will merge shortly

**Comment by rkooo567** [comment]
@franklsf95 Can you verify that after that PR is merged?

**Comment by rkooo567** [comment]
Let me assign P1 temporarily

**Comment by rkooo567** [comment]
@franklsf95 do you mind trying this again and see if you still see the issue? Eric's PR is merged now.

**Comment by franklsf95** [comment]
I just tried again. The C++ crash does not happen anymore. However, with 6GB object store memory (which is enough for one mapper output, so the tasks should be able to go through), the system deadlocks with the following output:

```
2021-04-06 16:36:29,597	INFO worker.py:653 -- Connecting to existing Ray cluster at address: 172.31.61.153:6379
[[ObjectRef(a67dc375e60ddd1affffffffffffffffffffffff0100000001000000)
  ObjectRef(a67dc375e60ddd1affffffffffffffffffffffff0100000002000000)
  ObjectRef(a67dc375e60ddd1affffffffffffffffffffffff0100000003000000)
  ObjectRef(a67dc375e60ddd1affffffffffffffffffffffff0100000004000000)]
 [ObjectRef(63964fa4841d4a2effffffffffffffffffffffff0100000001000000)
  ObjectRef(63964fa4841d4a2effffffffffffffffffffffff0100000002000000)
  ObjectRef(63964fa4841d4a2effffffffffffffffffffffff0100000003000000)
  ObjectRef(63964fa4841d4a2effffffffffffffffffffffff0100000004000000)]
 [ObjectRef(69a6825d641b4613ffffffffffffffffffffffff0100000001000000)
  ObjectRef(69a6825d641b4613ffffffffffffffffffffffff0100000002000000)
  ObjectRef(69a6825d641b4613ffffffffffffffffffffffff0100000003000000)
  ObjectRef(69a6825d641b4613ffffffffffffffffffffffff0100000004000000)]
 [ObjectRef(ee4e90da584ab0ebffffffffffffffffffffffff0100000001000000)
  ObjectRef(ee4e90da584ab0ebffffffffffffffffffffffff0100000002000000)
  ObjectRef(ee4e90da584ab0ebffffffffffffffffffffffff0100000003000000)
  ObjectRef(ee4e90da584ab0ebffffffffffffffffffffffff0100000004000000)]]
[2m[36m(pid=10659)[0m Mapper 0
[2m[36m(pid=10788)[0m Mapper 3
[2m[36m(pid=10795)[0m Mapper 1
[2m[36m(pid=10796)[0m Mapper 2
2021-04-06 16:36:42,357	WARNING worker.py:1083 -- Traceback (most recent call last):
  File "python/ray/_raylet.pyx", line 713, in ray._raylet.restore_spilled_objects_handler
  File "python/ray/_raylet.pyx", line 716, in ray._raylet.restore_spilled_objects_handler
  File "/home/ubuntu/miniconda3/envs/raysort/lib/python3.8/site-packages/ray/external_storage.py", line 487, in restore_spilled_objects
    return _external_storage.restore_spilled_objects(object_refs,
  File "/home/ubuntu/miniconda3/envs/raysort/lib/python3.8/site-packages/ray/external_storage.py", line 303, in restore_spilled_objects
    self._put_object_to_store(metadata, buf_len, f, object_ref,
  File "/home/ubuntu/miniconda3/envs/raysort/lib/python3.8/site-packages/ray/external_storage.py", line 100, in _put_object_to_store
    worker.core_worker.put_file_like_object(metadata, data_size, file_like,
  File "python/ray/_raylet.pyx", line 1053, in ray._raylet.CoreWorker.put_file_like_object
  File "python/ray/_raylet.pyx", line 1019, in ray._raylet.CoreWorker._create_put_buffer
  File "python/ray/_raylet.pyx", line 148, in ray._raylet.check_status
ray.exceptions.ObjectStoreFullError: Failed to put object a67dc375e60ddd1affffffffffffffffffffffff0100000001000000 in object store because it is full. Object size is 1073742074 bytes.
The local object store is full of objects that are still in scope and cannot be evicted. Tip: Use the `ray memory` command to list active objects in the cluster.
An unexpected internal error occurred while the IO worker was restoring spilled objects.
2021-04-06 16:36:42,357	WARNING worker.py:1083 -- Traceback (most recent call last):
  File "python/ray/_raylet.pyx", line 713, in ray._raylet.restore_spilled_objects_handler
  File "python/ray/_raylet.pyx", line 716, in ray._raylet.restore_spilled_objects_handler
  File "/home/ubuntu/miniconda3/envs/raysort/lib/python3.8/site-packages/ray/external_storage.py", line 487, in restore_spilled_objects
    return _external_storage.restore_spilled_objects(object_refs,
  File "/home/ubuntu/miniconda3/envs/raysort/lib/python3.8/site-packages/ray/external_storage.py", line 303, in restore_spilled_objects
    self._put_object_to_store(metadata, buf_len, f, object_ref,
  File "/home/ubuntu/miniconda3/envs/raysort/lib/python3.8/site-packages/ray/external_storage.py", line 100, in _put_object_to_store
    worker.core_worker.put_file_like_object(metadata, data_size, file_like,
  File "python/ray/_raylet.pyx", line 1053, in ray._raylet.CoreWorker.put_file_like_object
  File "python/ray/_raylet.pyx", line 1019, in ray._raylet.CoreWorker._create_put_buffer
  File "python/ray/_raylet.pyx", line 148, in ray._raylet.check_status
ray.exceptions.ObjectStoreFullError: Failed to put object a67dc375e60ddd1affffffffffffffffffffffff0100000002000000 in object store because it is full. Object size is 1073742074 bytes.
The local object store is full of objects that are still in scope and cannot be evicted. Tip: Use the `ray memory` command to list active objects in the cluster.
An unexpected internal error occurred while the IO worker was restoring spilled objects.
2021-04-06 16:36:42,357	WARNING worker.py:1083 -- Traceback (most recent call last):
  File "python/ray/_raylet.pyx", line 713, in ray._raylet.restore_spilled_objects_handler
  File "python/ray/_raylet.pyx", line 716, in ray._raylet.restore_spilled_objects_handler
  File "/home/ubuntu/miniconda3/envs/raysort/lib/python3.8/site-packages/ray/external_storage.py", line 487, in restore_spilled_objects
    return _external_storage.restore_spilled_objects(object_refs,
  File "/home/ubuntu/miniconda3/envs/raysort/lib/python3.8/site-packages/ray/external_storage.py", line 303, in restore_spilled_objects
    self._put_object_to_store(metadata, buf_len, f, object_ref,
  File "/home/ubuntu/miniconda3/envs/raysort/lib/python3.8/site-packages/ray/external_storage.py", line 100, in _put_object_to_store
    worker.core_worker.put_file_like_object(metadata, data_size, file_like,
  File "python/ray/_raylet.pyx", line 1053, in ray._raylet.CoreWorker.put_file_like_object
  File "python/ray/_raylet.pyx", line 1019, in ray._raylet.CoreWorker._create_put_buffer
  File "python/ray/_raylet.pyx", line 148, in ray._raylet.check_status
ray.exceptions.ObjectStoreFullError: Failed to put object a67dc375e60ddd1affffffffffffffffffffffff0100000003000000 in object store because it is full. Object size is 1073742074 bytes.
The local object store is full of objects that are still in scope and cannot be evicted. Tip: Use the `ray memory` command to list active objects in the cluster.
An unexpected internal error occurred while the IO worker was restoring spilled objects.

(this last error repeats)
```

When I use a larger object store memory size, e.g. 8GB, the program finishes normally.

**Comment by franklsf95** [comment]
Fixed by #15757
